#22 Reflections over DynamicContract

jcarraror opened the pull request:
Function registering for dynamic contracts now depends on reflection.

We need to register the contract class in the contract header:

```c++
 /**
   * Register contract class via ContractReflectionInterface.
   */
  static void registerContract() {
    ContractReflectionInterface::registerContract<
        ERC20, const std::string &, const std::string &, const uint8_t &,
        const uint256_t &, ContractManager::ContractManagerInterface &,
        const Address &, const Address &, const uint64_t &,
        const std::unique_ptr<DB> &>(
        std::vector<std::string>{"erc20_name", "erc20_symbol", "erc20_decimals",
                                 "mintValue"},
        std::make_tuple("name", &ERC20::name, "view",
                        std::vector<std::string>{}),
        std::make_tuple("symbol", &ERC20::symbol, "view",
                        std::vector<std::string>{}),
        std::make_tuple("decimals", &ERC20::decimals, "view",
                        std::vector<std::string>{}),
        std::make_tuple("totalSupply", &ERC20::totalSupply, "view",
                        std::vector<std::string>{}),
        std::make_tuple("balanceOf", &ERC20::balanceOf, "view",
                        std::vector<std::string>{"_owner"}),
        std::make_tuple("transfer", &ERC20::transfer, "nonpayable",
                        std::vector<std::string>{"_to", "_value"}),
        std::make_tuple("approve", &ERC20::approve, "nonpayable",
                        std::vector<std::string>{"_spender", "_value"}),
        std::make_tuple("allowance", &ERC20::allowance, "view",
                        std::vector<std::string>{"_owner", "_spender"}),
        std::make_tuple("transferFrom", &ERC20::transferFrom, "nonpayable",
                        std::vector<std::string>{"_from", "_to", "_value"}));
  }
  ```
  
  And after this, we can change the registerContractFunctions overload from this:
  
  ```c++
  void ERC20::registerContractFunctions() {
  this->registerViewFunction(Hex::toBytes("0x06fdde03"), [this](const ethCallInfo &callInfo) {
    return this->name();
  });
  this->registerViewFunction(Hex::toBytes("0x95d89b41"), [this](const ethCallInfo &callInfo) {
    return this->symbol();
  });
  this->registerViewFunction(Hex::toBytes("0x313ce567"), [this](const ethCallInfo &callInfo) {
    return this->decimals();
  });
  this->registerViewFunction(Hex::toBytes("0x18160ddd"), [this](const ethCallInfo &callInfo) {
    return this->totalSupply();
  });
  this->registerViewFunction(Hex::toBytes("0x70a08231"), [this](const ethCallInfo &callInfo) {
    std::vector<ABI::Types> types = { ABI::Types::address };
    ABI::Decoder decoder(types, std::get<5>(callInfo).substr(4));
    return this->balanceOf(decoder.getData<Address>(0));
  });
  this->registerViewFunction(Hex::toBytes("0xdd62ed3e"), [this](const ethCallInfo &callInfo) {
    std::vector<ABI::Types> types = { ABI::Types::address, ABI::Types::address };
    ABI::Decoder decoder(types, std::get<5>(callInfo).substr(4));
    return this->allowance(decoder.getData<Address>(0), decoder.getData<Address>(1));
  });

  this->registerFunction(Hex::toBytes("0xa9059cbb"), [this](const ethCallInfo &callInfo) {
    std::vector<ABI::Types> types = { ABI::Types::address, ABI::Types::uint256 };
    ABI::Decoder decoder(types, std::get<5>(callInfo).substr(4));
    this->transfer(decoder.getData<Address>(0), decoder.getData<uint256_t>(1));
  });
  this->registerFunction(Hex::toBytes("0x095ea7b3"), [this](const ethCallInfo &callInfo) {
    std::vector<ABI::Types> types = { ABI::Types::address, ABI::Types::uint256 };
    ABI::Decoder decoder(types, std::get<5>(callInfo).substr(4));
    this->approve(decoder.getData<Address>(0), decoder.getData<uint256_t>(1));
  });
  this->registerFunction(Hex::toBytes("0x23b872dd"), [this](const ethCallInfo &callInfo) {
    std::vector<ABI::Types> types = { ABI::Types::address, ABI::Types::address, ABI::Types::uint256 };
    ABI::Decoder decoder(types, std::get<5>(callInfo).substr(4));
    this->transferFrom(decoder.getData<Address>(0), decoder.getData<Address>(1), decoder.getData<uint256_t>(2));
  });
}
```

To this:

```c++
void ERC20::registerContractFunctions() {
  registerContract();
  this->registerMemberFunction("name", &ERC20::name, this);
  this->registerMemberFunction("symbol", &ERC20::symbol, this);
  this->registerMemberFunction("decimals", &ERC20::decimals, this);
  this->registerMemberFunction("totalSupply", &ERC20::totalSupply, this);
  this->registerMemberFunction("balanceOf", &ERC20::balanceOf, this);
  this->registerMemberFunction("allowance", &ERC20::allowance, this);
  this->registerMemberFunction("transfer", &ERC20::transfer, this);
  this->registerMemberFunction("approve", &ERC20::approve, this);
  this->registerMemberFunction("transferFrom", &ERC20::transferFrom, this);
}
```

So, now the developer don't need to pass the contract signature neither use a proper function for view, payable, etc.

Jean-Lessa commented on src/contract/abi.h:
```diff
@@ -2,263 +2,306 @@
+/// Class that encodes and packs native data types into Solidity ABI strings.
+class Encoder {
+private:
+  std::string data; ///< Encoded Solidity ABI string, as RAW BYTES. Use
```
Doxygen might not understand the comment line split like that IIRC. You can either leave it all in one contiguous line:

```c++
std::string data; ///< Encoded Solidity ABI string, as RAW BYTES. Use Hex::fromBytes().get() to print it properly.
```

Or bring the comment to above the line:

```c++
/**
 * Encoded Solidity ABI string, as RAW BYTES.
 * Use Hex::fromBytes().get() to print it properly.
 */
std::string data;
```

Your call, but I suggest the latter since it's two lines anyway.

jcarraror replied:
Done!

Jean-Lessa commented on src/contract/abi.h:
```diff
@@ -2,263 +2,306 @@
-       */
-      Encoder(const ABI::Encoder::EncVar& data, std::string_view func = "");
+  /// Getter for `data`.
+  /// TODO: rename this to getData() or something similar (for consistency with
```
Ita's probably working on this in another branch? If yes please ignore this comment.

Jean-Lessa commented on src/contract/abi.h:
```diff
@@ -2,263 +2,306 @@
-        if (std::holds_alternative<T>(this->data[index])) return std::get<T>(this->data[index]);
-        throw std::runtime_error("Type mismatch");
-      }
+  std::any
```
Missing Doxygen docs.

jcarraror replied:
Done!

Jean-Lessa commented on src/contract/contractmanager.cpp:
```diff
@@ -1,62 +1,78 @@
-  rdpos(rdpos),
-  options(options),
-  interface(*this) {
+ContractManager::ContractManager(State *state, const std::unique_ptr<DB> &db,
```
Formatting is a little wonky here, but the content seems to be the same.

Jean-Lessa commented on src/contract/contractmanager.cpp:
Seems like (almost) the whole file has the same contents, just the formatting seems a bit wonky on some places. This is fine, we can fix those later.

Jean-Lessa commented on src/contract/erc20.cpp:
Same as contractmanager.cpp.

jcarraror replied:
Done!

Jean-Lessa commented on src/utils/contractreflectioninterface.h:
```diff
@@ -0,0 +1,917 @@
+}
+
+static std::unordered_map<std::string, std::vector<std::string>>
+    constructorArgumentNamesMap; /// Map to store constructor argument names
```
Doxygen might not understand the line comments like that, append an "<" (e.g. `///< Map to...`).

jcarraror replied:
Done!

Jean-Lessa reviewed (approved):
Seems all good from an initial look. Left a few minor thoughts in the comments.

jcarraror commented on src/utils/contractreflectioninterface.h:
I agree. Maybe all the ABI templates handlers need to be moved to ABI namespace. Also, tmaybe he json object creation needs a proper file. The refl interface it's open to a refactor, of course.

itamarcps commented:
Within commit 3113efc71082de4848ed57b845d6dd3c14934cd6, I have fixed the compiler issues that the merge 746cf805548b6807cf618189f07c6ef6a31ce4d0 caused, The issues were:

- Circular dependency of `utils/contractreflectionsinterface.h` and `contract/dynamiccontract.h`, the circular dependency was due to the `registerCoreContractClasses` function on the interface using the `DynamicContract`.
   This issue was solved my moving the `registerCoreContractClasses` into a source file, and including the `dynamiccontract.h` only there
 
 - Duplicated functions and variables on `ContraftReflectionsInterface`
 The of the duplicated functions are: `registerContract() and isContractRegistered()`, and for the variables, they are `constructorArgumentNamesMap`, `methodMutabilityMap` and `argumentNamesMap`, for functions, I Kept the longest one and variables the ones declared in the top of the file.
 
 I have not checked if it is still working as expected, @jcarraror Please review my changes.

jcarraror commented:
> Within commit [3113efc](https://github.com/SparqNet/orbitersdk-cpp/commit/3113efc71082de4848ed57b845d6dd3c14934cd6), I have fixed the compiler issues that the merge [746cf80](https://github.com/SparqNet/orbitersdk-cpp/commit/746cf805548b6807cf618189f07c6ef6a31ce4d0) caused, The issues were:
> 
>     * Circular dependency of `utils/contractreflectionsinterface.h` and `contract/dynamiccontract.h`, the circular dependency was due to the `registerCoreContractClasses` function on the interface using the `DynamicContract`.
>       This issue was solved my moving the `registerCoreContractClasses` into a source file, and including the `dynamiccontract.h` only there
> 
>     * Duplicated functions and variables on `ContraftReflectionsInterface`
>       The duplicated functions are: `registerContract() and isContractRegistered()`, and the variables, they are `constructorArgumentNamesMap`, `methodMutabilityMap` and `argumentNamesMap`, for functions, I Kept the longest one and variables the ones declared in the top of the file.
> 
> 
> I have not checked if it is still working as expected, @jcarraror Please review my changes.

@itamarcps , the only thing missing was the inclusion of writeContractsToJson` implementation in the interface header. In addition, the `registerCoreContractClasses` it's deprecated, so i removed it from the source. Everything working fine now.
